### PR TITLE
Remove parent cells data structure from the code

### PR DESCRIPTION
**Pull request by sjplimp**
## Purpose

Previously, a copy of all parent cells within the hierarchical grid were stored by every processor.  For large grids with many levels of adaptation this can consume a lot of memory.  With this change, parent cells are no longer stored, only child cells.

## Author(s)

Steve

## Backward Compatibility

This alters the syntax for the write_grid and read_grid commands, as well as the format for the file they write/read.  It replaces the nparent and nchild stat_style keywords with ngrid.  It enhances the options for the create_grid command.

## Implementation Notes

_Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in SPARTA are affected_

## Post Submission Checklist

_Please check the fields below as they are completed_
- [ ] The feature or features in this pull request is complete
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] One or more example input decks are included
- [ ] The source code follows the SPARTA formatting guidelines

## Further Information, Files, and Links

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

### CODE REVIEW

**Comment by sjplimp** [comment]
Two features are not yet enabled by this PR:
(1) distributed explicit surfs - requires a new algorithm for finding grid/surf overlaps
(2) Kokkos - @stanmoore1 can you please look at this?

This should also be tested against all examples and regression tests we have to see if anything breaks.

**Comment by sjplimp** [comment]
Also, I need to update the doc pages to reflect this new version,
including some minor changes to command syntax.

**Comment by aborner1** [comment]
@sjplimp How much memory does it take to store (child) cells in this new implementation? Still around 300 Bytes/cell?

**Comment by sjplimp** [comment]
@aborner1 Essentially the same.  There is one extra int added to ChildCell, but that's in the noise.

**Comment by sjplimp** [comment]
@aborner1 What is different is that each parent cell can no longer have an arbitrary grid of child cells within it.  Now, all the parent cells at level N have to define the same grid of child cells at level N+1.  Create_grid and adapt_grid did not expose this option previously, but in principle it was possible.

**Comment by sjplimp** [comment]
@aborner1 Actually I was wrong about the memory.  There is no difference in ChildCell.  One int was removed and another added.

File: src/fix_ave_grid.cpp [hunk 1]
```diff
@@ -504,8 +504,9 @@ void FixAveGrid::end_of_step()
           k = umap[m][0];
           if (j == 0) {
             double *compute_vector = compute->vector_grid;
-            for (i = 0; i < nglocal; i++)
+            for (i = 0; i < nglocal; i++) {
               tally[i][k] += compute_vector[i];
+	    }
```
Comment (stanmoore1) [inline src/fix_ave_grid.cpp:509]: Looks like wrong indentation, maybe left over from debugging?

**Review by stanmoore1 — CHANGES_REQUESTED** [review]

**Comment by stanmoore1** [comment]
Kokkos now compiles, I'm running this through full Kokkos regression testing.

**Comment by stanmoore1** [comment]
@sjplimp I'm seeing a segfault with FFTs:

```
Program received signal SIGSEGV, Segmentation fault.
0x00000000004dcbf5 in SPARTA_NS::ComputeFFTGrid::reallocate (
    this=this@entry=0x11642f0) at ../compute_fft_grid.cpp:614
614	      array_grid[map2[i]][icol] = gridwork[i];
Missing separate debuginfos, use: debuginfo-install glibc-2.17-317.el7.x86_64 libibverbs-22.4-5.el7.x86_64 libnl3-3.2.28-4.el7.x86_64 librdmacm-22.4-5.el7.x86_64 sssd-client-1.16.5-10.el7_9.5.x86_64
(gdb) where
#0  0x00000000004dcbf5 in SPARTA_NS::ComputeFFTGrid::reallocate (
    this=this@entry=0x11642f0) at ../compute_fft_grid.cpp:614
#1  0x00000000004dd43a in SPARTA_NS::ComputeFFTGrid::ComputeFFTGrid (
    this=0x11642f0, sparta=<optimized out>, narg=13, arg=0xc6cd20)
    at ../compute_fft_grid.cpp:261
#2  0x0000000000569e77 in SPARTA_NS::Modify::add_compute (this=0xc687a0, 
    narg=13, arg=0xc6cd20) at ../compute_fft_grid.h:17
#3  0x000000000048f3ed in compute (this=0xc47cc0) at ../input.cpp:1245
#4  SPARTA_NS::Input::execute_command (this=this@entry=0xc47cc0)
    at ../input.cpp:673
#5  0x000000000048fea3 in SPARTA_NS::Input::file (this=0xc47cc0)
    at ../input.cpp:199
#6  0x0000000000406386 in main (argc=<optimized out>, argv=<optimized out>)
    at ../main.cpp:30
```

**Comment by stanmoore1** [comment]
I'm seeing a few failures with the `adapt` examples and Kokkos, looking into it now.

**Comment by stanmoore1** [comment]
The Kokkos issue is now fixed.

File: src/grid.cpp [hunk 2]
```diff
@@ -812,281 +805,161 @@ void Grid::rehash()
...
-        } else {
-          neigh[ZHI] = -index-1;
-          if (boundary) nmask = neigh_encode(NPBPARENT,nmask,ZHI);
-          else nmask = neigh_encode(NPARENT,nmask,ZHI);
-        }
-      }
+      if (icell < nlocal) unknownflag = 1;
     }
-
+
```
Comment (stanmoore1) [inline src/grid.cpp:937]: Whitespace issue

File: src/fix_ave_grid.cpp (hunk 1 above)
Comment (stanmoore1) [reply src/fix_ave_grid.cpp]: @sjplimp

**Review by stanmoore1 — APPROVED** [review]
@sjplimp I think we should merge this when ready and fix whitespace issues in a separate PR.

**Comment by sjplimp** [comment]
@stanmoore1 if this version passes all regression tests, it can be merged

**Review by stanmoore1 — APPROVED** [review]

**Pull request merged**